File: Limbus/limbus_core/frontend/Scanner.fs
module Scanner
open System.Runtime.CompilerServices
open Token
open Source

[<AbstractClass>]
type Scanner(source : Source) =
    class
        let mutable current_token : Token = EodOfToken  (source) :> Token
    
        member this.source = source
        
        abstract member extra_token : unit -> Token

        member this.cuurent_token() : Token = 
            current_token
            
        member this.next_token() : Token = 
            current_token <- this.extra_token()
            current_token
            
        member this.currnet_char() : TCHAR =
            this.source.current_char()
            
        member this.next_char() : TCHAR =
            this.source.next_char()
    
    end 

File: Limbus/limbus_core/frontend/scanner.py
# -*- coding: utf-8 -*-

from abc import ABCMeta, abstractmethod


class Scanner(metaclass=ABCMeta):
    EOF = None

    def __init__(self, source):
        self.source = source
        self.cur_token = None

    def current_token(self):
        return self.cur_token

    def next_token(self):
        self.cur_token = self.extract_token()
        return self.cur_token

    @abstractmethod
    def extract_token(self):
        raise NotImplementedError()

    def current_char(self):
        return self.source.current_char()

    def next_char(self):
        return self.source.next_char()
